File: 0846_handOfStraights/main.cpp
#include <vector>
#include <map>
#include <queue>

class Solution {
public:
    bool isNStraightHand(std::vector<int>& hand, int groupSize) {
        if(hand.size() % groupSize !=0)
            return false;
        
        std::map<int, int> myMap;
        for(int i = 0; i < hand.size(); ++i) {
            ++myMap[hand[i]];
        }

        while(myMap.size() != 0) {
            int cur = myMap.begin()->first;
            for(int i = 0; i < groupSize; ++i) {
                if(myMap[cur+i] == 0)
                    return false;
                else if(--myMap[cur+i] == 0)
                    myMap.erase(cur+i);
            }
        }
        return true;

    }
};

File: 0846_handOfStraights/main.py
from typing import List

class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        if len(hand) % groupSize:
            return False

        # map
        count = {}
        for val in hand:
            count[val] = 1 + count.get(val, 0)

        # min heap
        minHeap = list(count.keys())
        heapq.heapify(minHeap)
        while minHeap:
            first = minHeap[0]
            for i in range(first, first + groupSize):
                if i not in count:
                    return False
                count[i] -= 1
                if count[i] == 0:
                    # the next round won't be able to form a group
                    if i != minHeap[0]:
                        return False
                    heapq.heappop(minHeap)
        return True
        